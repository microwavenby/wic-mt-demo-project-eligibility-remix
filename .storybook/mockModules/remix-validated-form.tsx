import { Form } from "@remix-run/react";
export {
  validationError,
  setFormDefaults,
  createValidator,
  FieldErrors,
  Validator,
} from "../../node_modules/remix-validated-form";
// This is duplicative of the jest helper remixValidatedFormMock,
// but the Jest global missing breaks the component in Storybook

// This ends up being <form> in ./remix-react.tsx
export const ValidatedForm = Form;

type MinimalInputProps = {
  onChange?: (...args: any[]) => void;
  onBlur?: (...args: any[]) => void;
  defaultValue?: any;
  defaultChecked?: boolean;
  name?: string;
  type?: string;
};

export const getInputProps = <T extends MinimalInputProps>(
  props = {} as any
) => {
  return props as T;
};

// TODO: Figure out a way to determine whether this component is being
// rendered in a story with "error" in the name or route, so that
// an example error can be shown in Storybook
export const useField = (args: any) => {
  return { error: undefined, getInputProps };
};
