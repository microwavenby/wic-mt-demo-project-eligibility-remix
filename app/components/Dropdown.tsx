import { useTranslation } from "react-i18next";
import { ChangeEvent, ReactElement } from "react";
import {
  Dropdown as USWDS_Dropdown,
  Label,
  ErrorMessage,
} from "@trussworks/react-uswds";

import Required from "app/components/Required";

import { i18nKey } from "app/types";
import { useField } from "remix-validated-form";

export interface DropdownProps {
  handleChange: (e: ChangeEvent<HTMLSelectElement>) => void;
  id: string;
  labelKey: i18nKey;
  options: string[];
  required?: boolean;
  selectedOption?: string;
}

// @TODO: This component expects pre-translated option strings.
//        It should be refactored if itos ever used with non-integer options.
const Dropdown = (props: DropdownProps): ReactElement => {
  const { handleChange, id, labelKey, options, required, selectedOption } =
    props;
  const { getInputProps, error } = useField(id);
  let { t } = useTranslation();
  return (
    <>
      <Label htmlFor={id}>
        {t(labelKey)}
        {required && <Required />}
      </Label>
      {error && (
        <ErrorMessage id="${titleKey}-error-message">{error}</ErrorMessage>
      )}
      <USWDS_Dropdown
        id={id}
        name={id}
        onChange={handleChange}
        value={selectedOption}
      >
        <option
          value=""
          {...getInputProps({ id: `${id}-unselected`, label: "" })}
        >
          -&nbsp;
          {t("select")}
          &nbsp;-
        </option>
        {options.map((option: string, index: number) => (
          <option
            key={option}
            {...getInputProps({
              id: `${id}-${index + 1}`,
              label: option,
              value: option,
            })}
          >
            {option}
          </option>
        ))}
      </USWDS_Dropdown>
    </>
  );
};

export default Dropdown;
