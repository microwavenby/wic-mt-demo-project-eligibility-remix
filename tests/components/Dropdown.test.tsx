import { render, screen } from "@testing-library/react";
import { renderWithRouter } from "tests/helpers/setup";
import Dropdown, { DropdownProps } from "app/components/Dropdown";
import { useField } from "remix-validated-form";
jest.mock("remix-validated-form");
import { mockUseFieldReturnValue } from "tests/helpers/remixValidatedFormMock";

const mockedUseField = jest.mocked(useField);
mockedUseField.mockReturnValue(mockUseFieldReturnValue);

const testProps: DropdownProps = {
  handleChange: (e) => {},
  id: "test-id",
  labelKey: "dropdown label",
  options: ["a", "b", "c"],
};

it("should match snapshot", () => {
  const { container } = renderWithRouter(<Dropdown {...testProps} />);
  expect(container).toMatchSnapshot();
});

it("should match display required marker if required is true", () => {
  renderWithRouter(<Dropdown {...testProps} required={true} />);
  const required = screen.getByText("*");
  expect(required).toBeInTheDocument;
});

it("should show the selected element if passed in", () => {
  renderWithRouter(<Dropdown {...testProps} selectedOption="c" />);
  const required = screen.getByRole("option", {
    name: "c",
  }) as HTMLOptionElement;
  expect(required.selected).toBe(true);
});
