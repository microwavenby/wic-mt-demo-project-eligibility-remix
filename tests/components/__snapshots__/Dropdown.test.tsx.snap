// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div>
  <label
    class="usa-label"
    data-testid="label"
    for="test-id"
  >
    dropdown label
  </label>
  <select
    class="usa-select"
    data-testid="dropdown"
    id="test-id"
    name="test-id"
  >
    <option
      id="test-id"
      label=""
      value=""
    >
      - 
      Select
       -
    </option>
    <option
      id="test-id"
      label="a"
      value="a"
    >
      a
    </option>
    <option
      id="test-id"
      label="b"
      value="b"
    >
      b
    </option>
    <option
      id="test-id"
      label="c"
      value="c"
    >
      c
    </option>
  </select>
</div>
`;
